### PR TITLE
Add Webservice tutorial in (Modules) extension concept section

### PR DESCRIPTION
**Pull request by marsaldev**
Add new page about how to extend the Webservice resources through custom entity and module

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.x
| Description?  | Chatting with a community member that needs to add informations accessible through webservice, I discover that there isn't a small tutorial on how to do through a module with a custom entity (using ObjectModel).
| Fixed ticket? | ~

### CODE REVIEW

File: modules/concepts/webservice/_index.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        'forbidden_method' => array('DELETE') // optional
+    )
+  ];
+}
+```
+
+### Composer configuration to use namespaces in your ObjectModel entity
+
+To correctly load your class (Entity) you need to configure correctly your composer.json
+The follow is an example of correct configuration to load your entities in whatever folder you want.
```
Comment (matthieu-rolland) [inline modules/concepts/webservice/_index.md]:
```suggestion
The following is an example of a correct configuration to load your entities in whatever folder you want.
```

File: modules/concepts/webservice/_index.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      'title' => array('required' => true),
+      'type' => array('required' => true),
+      'content' => array(),
+      'meta_title' => array(),
+    )
+  );
+}
+```
+
+The `$webserviceParameters` is mandatory to define how and which fields to expose.
```
Comment (matthieu-rolland) [inline modules/concepts/webservice/_index.md]:
```suggestion
The `$webserviceParameters` array is mandatory to define how and which fields to expose.
```

File: modules/concepts/webservice/_index.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+The name to access to all the elements:
```
Comment (matthieu-rolland) [inline modules/concepts/webservice/_index.md]:
```suggestion
The key to access all the elements:
```

File: modules/concepts/webservice/_index.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+```php
+'objectsNodeName' => 'articles'
+```
+
+The name to access to the single element:
```
Comment (matthieu-rolland) [inline modules/concepts/webservice/_index.md]:
```suggestion
The key to access a single element:
```

**Review by matthieu-rolland — COMMENTED** [review]
that's a great addition 👍 

I made some feedback which I think makes it easier to read

**Review by matks — COMMENTED** [review]
Looks good to me, one last check from @kpodemski who knows better than me the topic and let's merge

**Comment by kpodemski** [comment]
Thanks @marsaldev - I'll take a look this week ❤️

**Comment by Matt75** [comment]
Hello @marsaldev 

I'm not in favor to usage of namespace for classes that extends legacy PrestaShop Core classes like ObjectModel.

This classes were designed to be in the global namespace, use it with a namespace can create issues with others legacy classes, like native legacy logger : https://github.com/PrestaShop/PrestaShop/issues/28617

And also various others pieces of code that will interact with it in legacy PrestaShop core can encountered issues cause it were designed only for global namespace.

So my recommendation is to keep classes that's extends legacy PrestaShop core classes in global namespace, so put your ObjectModel in `modules/yourmodule/classes/Article.php`

All new classes that's not extends legacy classes can be namespaced and placed in `modules/yourmodule/src/` without issue.

**Comment by kpodemski** [comment]
Hi @marsaldev 

I agree with @Matt75. 

It's been reported a few times that having legacy classes hidden behind namespaces may be problematic in some undocumented scenarios. 

Let's keep it simple and focus on handling WS inside a class without having the composer handling autoloader, etc.

**Review by kpodemski — CHANGES_REQUESTED** [review]
Please drop the use of namespaces and composer

**Comment by marsaldev** [comment]
Got it! 😉
I'll update it 👍🏻
Thanks a lot @Matt75 @kpodemski

**Review by marsaldev — COMMENTED** [review]
Done 😊
Looking forward your feedback(s)

File: modules/concepts/webservice/_index.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+  protected $webserviceParameters = array(
+    'objectNodeName' => 'article',
+    'objectsNodeName' => 'articles',
+    'fields' => array(
+      'title' => array('required' => true),
+      'type' => array('required' => true),
+      'content' => array(),
+      'meta_title' => array(),
+    )
+  );
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
    protected $webserviceParameters = array(
      'objectNodeName' => 'article',
      'objectsNodeName' => 'articles',
      'fields' => array(
          'title' => array('required' => true),
          'type' => array('required' => true),
          'content' => array(),
          'meta_title' => array(),
        )
    );
```

File: modules/concepts/webservice/_index.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+```php
+'objectNodeName' => 'articles'
+```
+
+The parameter to set which fields to expose through the webservice and settings for each of them:
+```php
+'fields' => array()
+```
+
+### Register it through the addWebserviceResources hook
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
### Registration with a hook
```

File: modules/concepts/webservice/_index.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+The hook `addWebserviceResources` must be registered by your module, usually done during the module installation.
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
The hook `addWebserviceResources` must be registered and implemented by your module.
```

File: modules/concepts/webservice/_index.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+```php
+public function hookAddWebserviceResources($params)
+{
+  return [
+    'articles' => array(
+        'description' => 'Blog articles', // The description for who access to this resource through WS
+        'class' => 'Article', // The classname of your Entity
+        'forbidden_method' => array('DELETE') // optional if you want to forbid some methods
+    )
+  ];
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
    return [
      'articles' => [
          'description' => 'Blog articles', // The description for those who access to this resource through WS
          'class' => 'Article', // The classname of your Entity
          'forbidden_method' => array('DELETE') // Optional, if you want to forbid some methods
      ]
    ];
```

File: modules/concepts/webservice/_index.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+}
+```
+
+### Load your entity
+Don't forget to include the class file of your entity (i.e. Article.php) at the top of your main module file.
+The following is an example of a correct configuration to load your entities in whatever folder you want.
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
The following is an example of a correct configuration to load your example entity to your module.
```

File: modules/concepts/webservice/_index.md [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+```php
+include_once dirname(__FILE__).'/src/Entity/Article.php';
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
include_once _PS_MODULE_DIR_ . 'wsarticle/src/Entity/Article.php';
```

File: modules/concepts/webservice/_index.md [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+### Complete example of a main module file
+
+```php
+<?php
+
+if (!defined('_PS_VERSION_')) {
+    exit;
+}
+
+include_once dirname(__FILE__).'/src/Entity/Article.php';
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
include_once _PS_MODULE_DIR_ . 'wsarticle/src/Entity/Article.php';
```

File: modules/concepts/webservice/_index.md [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+            `date_upd` datetime NOT NULL,
+            PRIMARY KEY  (`id_article`)
+        ) ENGINE='._MYSQL_ENGINE_.' DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci';
+
+        $sql_lang = 'CREATE TABLE IF NOT EXISTS `'._DB_PREFIX_.Article::$definition['table'].'_lang` (
+            `id_article` int(10) unsigned NOT NULL,
+            `id_lang` int(10) unsigned NOT NULL,
+            `title` varchar(255),
+            `content` text NOT NULL,
+            `meta-title` varchar(255) NOT NULL,
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
            `meta_title` varchar(255) NOT NULL,
```

File: modules/concepts/webservice/_index.md [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        
+        return false;
+    }
+
+    public function hookAddWebserviceResources($params)
+    {
+        return [
+            'articles' => array(
+                'description' => 'Blog articles', // The description for who access to this resource through WS
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
                'description' => 'Blog articles', // The description for those who access to this resource through WS
```

File: modules/concepts/webservice/_index.md [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+                'class' => 'Article', // The classname of your Entity
+                'forbidden_method' => array('DELETE') // optional if you want to forbid some methods
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
                'forbidden_method' => array('DELETE') // Optional, if you want to forbid some methods
```

File: modules/concepts/webservice/_index.md [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+            )
+        ];
+    }
+}
+```
+
+### Final notes
+
+Following the example above, the new resource will be available in the webservices resources list:
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
Following the example above, the new resource will be available in the webservice resources list:
```

File: modules/concepts/webservice/_index.md [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+
+{{< figure src="../img/new-ws-resource.png" title="New Webservice resource" >}}
+
+And will be accessible through your api url, plus the name that you've decided in `objectsNodeName`, for instance:
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
This new resource will be accessible through your API endpoint behind the name that you set in `objectsNodeName`, for instance:
```

File: modules/concepts/webservice/_index.md [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+
+`https://mywebsite.shop/api/articles`
+
+will give you something similar(in browser):
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
will return something similar to what you can see on the screenshot below (it is the XML response previewed in the browser):
```

File: modules/concepts/webservice/_index.md [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+
+{{< figure src="../img/empty-articles.png" title="Webservice articles list (empty)" >}}
+
+And something similar when you have articles in database:
```
Comment (kpodemski) [inline modules/concepts/webservice/_index.md]:
```suggestion
And something like on the next screenshot, if you have articles in the database:
```

**Review by kpodemski — APPROVED** [review]

**Comment by kpodemski** [comment]
Thank you, @marsaldev ❤️ I allowed myself to commit some minor improvements. Let's wait for the build to see how it looks, if it is ok we will need to have it in the PrestaShop 8 too :)

**Pull request merged**